src/Terrain/MapWindow.java

```diff
@@ -191,12 +212,25 @@ public void revealFog(Unit scout, Path movepath)
     {
       for( PathNode node : movepath.getWaypoints() )
       {
-        for( XYCoord coord : Utils.findVisibleLocations(this, scout, node.x, node.y) )
+        for( XYCoord coord : Utils.findVisibleLocations(this, scout, node.x, node.y, false) )
         {
-          isFogged[coord.xCoord][coord.yCoord] = false;
-          lastOwnerSeen[coord.xCoord][coord.yCoord] = master.getLocation(coord).getOwner();
+          revealFog(coord, false);
+        }
+        // We need to do a second pass with piercing vision so we can know whether to reveal the units
```

Comment:
I'm not a fan of this double-pass approach. Seems like we should be able to just figure it out in one go.

Reply:
There doesn't seem to be an easy way to return both sets of info - the places you can reveal cloaked units, and the places you can only see into.

Reply:
Isn't the rule for revealing a cloaked unit the same as seeing units under cover in fog?

Reply:
It is, but whether the cover tile is "fogged" is equal to whether you can see a unit in it.
If a sub is sitting in open water, you'd be able to see the water perfectly well while not seeing the sub.

Also, yes, the rule is the same (and this is a deliberate departure from the carts; piercing vision doesn't help you there vs hidden units)